### PR TITLE
Triton FIL backend XGBoost notebook and README

### LINKED ISSUE #9: Provide example notebook
Create example notebook demonstrating usage with both http and grpc as well as both xgboost and lightgbm

### PR DESCRIPTION
**Pull request by abhisheksawarkar**
PR for merging notebook and readme into the main branch

Close #9

### CODE REVIEW

**Comment by wphicks** [comment]
okay to test

**Comment by wphicks** [comment]
Depends on resolution of #106 (PR should go up tomorrow) for `model_analyzer` portion

**Comment by wphicks** [comment]
CI failure is unrelated. Will rerun once CI is fixed

File: notebooks/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Trtion FIL backend with XGBoost
+
+This notebook is a reference for deploying a forest model trained using XGBoost library in Triton Inference Server with Forest Inference Library (FIL) backend. The notebook explains how one can deploy XGBoost model in Triton, check deployment status and send inference requests, set concurrent model execution and dynamic batching and find the best deployment configuration using Model Analyzer.
+
+## Requirements
+* Nvidia GPU (T4 or V100 or A100)
```
Comment (wphicks) [inline notebooks/README.md:6]: Let's make it clear that these are not the *only* GPUs that work for this, just the recommended ones.

File: notebooks/README.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+* [Latest NVIDIA driver](https://docs.nvidia.com/datacenter/tesla/tesla-installation-notes/index.html)
+* [Docker](https://docs.docker.com/get-docker/)
+* [The NVIDIA container toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#docker)
+
+## Run the Triton Inference Server container 
+
+To run this notebook, pull the Triton container as mentioned in the README of [Triton Inference Server FIL backend](https://github.com/triton-inference-server/fil_backend) Github repo. Before running the container, clone the repository and then run the container:
```
Comment (wphicks) [inline notebooks/README.md:13]:
```suggestion
To run this notebook, pull the Triton container as mentioned in the README of [Triton Inference Server FIL backend](https://github.com/triton-inference-server/fil_backend#pre-built-container) Github repo. Before running the container, clone the repository and then run the container:
```

File: notebooks/README.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  --rm \
+  --net=host \
+  -v $PWD:/data \
+  nvcr.io/nvidia/tritonserver:<tag>    # Put the appropriate tag here
+  
+```
+
+## Install Jupyter notebook inside the Triton container
+```
+pip3 install jupyter notebook
```
Comment (wphicks) [inline notebooks/README.md:31]: Are both of these packages necessary?

File: notebooks/README.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+```
+git clone https://github.com/triton-inference-server/fil_backend.git
+cd fil_backend/notebooks
+
+docker run \
+  - it \
+  --gpus=all \
+  --rm \
+  --net=host \
```
Comment (wphicks) [inline notebooks/README.md:23]: Can we just expose the correct ports rather than using `--net=host`?

**Review by wphicks — CHANGES_REQUESTED** [review]
Just reviewed the README; I'll follow up with a review of the notebook after I've set up ReviewNB

File: notebooks/simple_xgboost_example.ipynb [hunk 5]
```diff
@@ -0,0 +1,714 @@
...
+   "source": [
+    "# Generate dummy data to perform binary classification\n",
+    "seed = 7\n",
+    "features = 9 # number of sample features\n",
+    "samples = 10000 # number of samples\n",
+    "X = numpy.random.rand(samples, features).astype('float32')\n",
+    "Y = numpy.random.randint(2, size=samples)\n",
+    "\n",
+    "test_size = 0.33\n",
+    "X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=test_size, random_state=7)"
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:141]:
```suggestion
    "X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=test_size, random_state=seed)"
```

File: notebooks/simple_xgboost_example.ipynb [hunk 6]
```diff
@@ -0,0 +1,714 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "y_pred = model.predict(X_test)\n",
+    "predictions = [round(value) for value in y_pred]\n",
+    "accuracy = accuracy_score(y_test, predictions)\n",
+    "print(\"Test Accuracy: %.2f%%\" % (accuracy * 100.0))"
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:163]: Prefer `.format` or f-strings to old-style string formatting

File: notebooks/simple_xgboost_example.ipynb [hunk 7]
```diff
@@ -0,0 +1,714 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Export and load XGBoost model in Triton Inference Server\n",
+    "\n",
+    "For deploying the trained XGBoost model in Triton Inference Server, follow the steps below:\n",
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:172]: Just to help underscore that the name is arbitrary, why don't we name this model `xgboost_example` and update the directory name accordingly?

File: notebooks/simple_xgboost_example.ipynb [hunk 8]
```diff
@@ -0,0 +1,714 @@
...
+    "model.save_model('/data/model_repository/fil/1/xgboost.model')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**2. Create and save config.pbtxt**\n",
+    "\n",
+    "To deploy the model in Triton Inference Server, we need to create and save a protobuf config file called config.pbtxt under `model_repository/fil/` directory that contains information about the model and the deployment. Sample config file is available here: [link](https://github.com/triton-inference-server/fil_backend#configuration)\n",
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:206]:
```suggestion
    "To deploy the model in Triton Inference Server, we need to create and save a protobuf config file called config.pbtxt under `model_repository/fil/` directory that contains information about the model and the deployment. Sample config file is available [here](https://github.com/triton-inference-server/fil_backend#configuration).\n",
```

File: notebooks/simple_xgboost_example.ipynb [hunk 9]
```diff
@@ -0,0 +1,714 @@
...
+    "    value: { string_value: \"0\" }\n",
+    "  }\n",
+    "]\n",
+    "\n",
+    "dynamic_batching {\n",
+    "  preferred_batch_size: [1024, 8192]\n",
+    "  max_queue_delay_microseconds: 100\n",
+    "}\n",
+    "```\n",
+    "Store the above config at `/data/model_repository/fil/` directory as config.pbtxt as follows:\n",
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:265]: Let's generate this file for them (either in the notebook or externally) so they don't have to go through the hassle of figuring that out inside the Docker container or thinking about how the bind-mount works.

File: notebooks/simple_xgboost_example.ipynb [hunk 10]
```diff
@@ -0,0 +1,714 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**2. Create and save config.pbtxt**\n",
+    "\n",
+    "To deploy the model in Triton Inference Server, we need to create and save a protobuf config file called config.pbtxt under `model_repository/fil/` directory that contains information about the model and the deployment. Sample config file is available here: [link](https://github.com/triton-inference-server/fil_backend#configuration)\n",
+    "\n",
+    "Essentially, the following parameters need to be updated as per your configuration\n",
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:208]: Let's try to signpost this a bit better. Tell them that we are going to generate the config file for them with the following command and then tell them that we are presenting the config here so that they know what they'll need to tweak if deploying a model of their own. Also, avoid "essentially." We should either be direct about what we're leaving out of discussion or be sure to include all relevant details.

File: notebooks/simple_xgboost_example.ipynb (hunk 9 above)
Comment (wphicks) [reply notebooks/simple_xgboost_example.ipynb:265]: Ah, I see. It's happening in the following cell. See my above comment about signposting.

File: notebooks/simple_xgboost_example.ipynb [hunk 11]
```diff
@@ -0,0 +1,714 @@
...
+    "  max_queue_delay_microseconds: 100\n",
+    "}\n",
+    "EOL"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Finally, the model repository should look like this:\n",
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:339]: "Finally" duplicated in both this and next line

File: notebooks/simple_xgboost_example.ipynb [hunk 12]
```diff
@@ -0,0 +1,714 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Run the Triton Inference Server in a Subprocess from Jupyter notebook\n",
+    "\n",
+    "# The os.setsid() is passed in the argument preexec_fn so\n",
+    "# it's run after the fork() and before  exec() to run the shell.\n",
+    "pro = subprocess.Popen([\"tritonserver --model-repository=/data/model_repository\"], stdout=subprocess.PIPE, \n",
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:364]: Do we need `shell=True`? If not, it should be avoided.

File: notebooks/simple_xgboost_example.ipynb [hunk 13]
```diff
@@ -0,0 +1,714 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Stopping Triton Server before proceeding further\n",
+    "os.killpg(os.getpgid(pro.pid), signal.SIGTERM)  # Send the signal to all the process groups"
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:401]: Is this necessary? Does `pro.kill()` not suffice?

File: notebooks/simple_xgboost_example.ipynb (hunk 12 above)
Comment (wphicks) [reply notebooks/simple_xgboost_example.ipynb:364]: Also, it might be nice to have a more descriptive variable name than `pro` here.

File: notebooks/simple_xgboost_example.ipynb [hunk 14]
```diff
@@ -0,0 +1,714 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Concurrent Model Execution\n",
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:408]: We should probably signpost this a bit better as well. Maybe it would be worth some sort of outline before we launch into everything explaining that we will be bringing the server up and down several times with different configuration options to demonstrate different features of the server? At the very least, while the server is up for the first time, we should fire off an inference request to demonstrate that it is a functional deployment.

File: notebooks/simple_xgboost_example.ipynb [hunk 15]
```diff
@@ -0,0 +1,714 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!export DCGM_VERSION=2.0.13\n",
+    "!wget -q https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/datacenter-gpu-manager_${DCGM_VERSION}_amd64.deb && \\\n",
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:583]: I thought DCGM was already in the Triton image? You're probably right that it's not, but I just want to make sure we're not installing this unnecessarily.

File: notebooks/simple_xgboost_example.ipynb [hunk 16]
```diff
@@ -0,0 +1,714 @@
...
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.8.10"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 4
+}
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:714]: Let's finish off with some sort of conclusion that clarifies what all they should have gotten out of this notebook.

**Review by wphicks — CHANGES_REQUESTED** [review]
Thanks so  much for all the hard work on this, @abhisheksawarkar! I've added some suggestions and requested some changes. The biggest thing for me overall is that it would be helpful to have some clearer high-level organization to this notebook, preferably organization that reflects how end-users will actually make use of this backend. Right now, the notebook appears to be structured around different components of Triton. That makes a certain amount of sense, but no user is going to want to e.g. set up the server without making use of the client. 

Maybe we can reorganize this with a clearer overall thrust toward the end-user's workflow? E.g. Train the model, start the server, demonstrate that the client works, show what current throughput is (via perf_analyzer), demonstrate that concurrent model execution improves throughput, use model_analyzer to determine optimal config parameters, update the config, demonstrate that we have improved throughput, finish up with a summary of how they would apply these steps to their own model or something like that.

File: notebooks/README.md (hunk 1 above)
Comment (dantegd) [reply notebooks/README.md:6]: It would be best to specify the GPU architecture (i.e. Pascal+ or Volta+) as well as recommended GPUs

File: notebooks/simple_xgboost_example.ipynb [hunk 17]
```diff
@@ -0,0 +1,714 @@
...
+    "# Triton server with FIL backend"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Overview\n",
+    "\n",
+    "This notebook will run through the procedure to deploy a XGBoost model in Triton Inference Server with Forest Inference Library (FIL) backend. The FIL backend allows forest models trained by several popular machine learning frameworks (including XGBoost, LightGBM, Scikit-Learn, and cuML) to be deployed in a Triton inference server using the RAPIDS Forest Inference LIbrary for fast GPU-based inference. Using this backend, forest models can be deployed seamlessly alongside deep learning models for fast, unified inference pipelines."
```
Comment (dantegd) [inline notebooks/simple_xgboost_example.ipynb:45]:
```suggestion
    "This notebook shows the procedure to deploy a XGBoost model in Triton Inference Server with Forest Inference Library (FIL) backend. The FIL backend allows forest models trained by several popular machine learning frameworks (including XGBoost, LightGBM, Scikit-Learn, and cuML) to be deployed in a Triton inference server using the RAPIDS Forest Inference LIbrary for fast GPU-based inference. Using this backend, forest models can be deployed seamlessly alongside deep learning models for fast, unified inference pipelines."
```

Comment (dantegd) [reply notebooks/simple_xgboost_example.ipynb:45]: As @wphicks mentioned at some point below if I'm mistaken, an outline of the notebook here would be quite nice and improve the discoverability of the notebook that is not a short one

File: notebooks/simple_xgboost_example.ipynb [hunk 18]
```diff
@@ -0,0 +1,714 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Run the Triton Inference Server with 2 instances of our model in a Subprocess from Jupyter notebook\n",
+    "\n",
+    "# The os.setsid() is passed in the argument preexec_fn so\n",
+    "# it's run after the fork() and before  exec() to run the shell.\n",
+    "pro = subprocess.Popen([\"tritonserver --model-repository=/data/model_repository\"], stdout=subprocess.PIPE, \n",
+    "                       shell=True, preexec_fn=os.setsid) "
```
Comment (dantegd) [inline notebooks/simple_xgboost_example.ipynb:506]: Another instance of the `pro` variable that could be a bit longer to be more descriptive and look nicer

File: notebooks/README.md (hunk 3 above)
Comment (abhisheksawarkar) [reply notebooks/README.md:31]: Jupyter notebook is not installed by default in the triton container. We'll need it to run the notebook in docker environment

Comment (wphicks) [reply notebooks/README.md:31]: Yep, but I think only `notebook` is sufficient to run the notebook, and I believe that is the recommended package for Jupyter installs

File: notebooks/simple_xgboost_example.ipynb (hunk 12 above)
Comment (abhisheksawarkar) [reply notebooks/simple_xgboost_example.ipynb:364]: Looks like `shell=True` is required or else the command throws an error.

**Review by wphicks — CHANGES_REQUESTED** [review]
Good! I think this looks much clearer now. It looks like you may have missed a couple of my earlier comments, so check over those again before finalizing. The biggest overall remaining issue that I see is making sure that you cover how to deal with permissions for files that pollute the host.

Also, don't forget to remove the checkpoints directory.

File: notebooks/.gitignore [hunk 19]
```diff
@@ -0,0 +1 @@
+.ipynb_checkpoints*
```
Comment (wphicks) [inline notebooks/.gitignore:1]: This can be added to the top-level .gitignore rather than introducing a second one.

File: notebooks/README.md [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+## Requirements
+* Nvidia GPU (Pascal+ Recommended GPUs: T4, V100 or A100)
+* [Latest NVIDIA driver](https://docs.nvidia.com/datacenter/tesla/tesla-installation-notes/index.html)
+* [Docker](https://docs.docker.com/get-docker/)
+* [The NVIDIA container toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#docker)
+
+## Run the Triton Inference Server container 
+
+**Note:** Please build the container locally inorder to use `model_analyzer` until the 21.08 release.
```
Comment (wphicks) [inline notebooks/README.md:13]:
```suggestion
**Note:** Due to a bug in release 21.07, Triton's `model_analyzer` cannot be used with the FIL backend until 21.08 is finalized. **Until then, you will need to build locally rather than pulling a pre-built container.** In the meantime, please follow [the instructions](https://github.com/triton-inference-server/fil_backend#building-locally) in the top-level README to build a development version of the 21.08 release.
```

Comment (wphicks) [reply notebooks/README.md:13]: Also, please link to the relevant section of the README.

File: notebooks/README.md [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+
+docker run \
+  -it \
+  --gpus=all \
+  --rm \
+  --net=host \
+  -v $PWD:/notebook \                    
+  nvcr.io/nvidia/tritonserver:<tag>  # Put the appropriate tag here.  
+```
+**Note:** The artifacts created by scripts inside the container are created with root permission. The user on host machine might not be able to modify the artifacts once the container exists. To avoid this issue, either make sure you have `sudo` access on host machine or copy the notebook `docker cp simple_xgboost_example.ipynb <docker_ID>` and create the artifacts inside the container.
```
Comment (wphicks) [inline notebooks/README.md:29]: Let's do one or the other here. It makes more sense to do the `docker cp` than fiddle with the permissions issues, so let's stick with that. Eliminate the bindmount and make the instructions for the `cp` clear and consistent with other instructions in this README.

**Review by wphicks — CHANGES_REQUESTED** [review]
Initial review of recent changes; more to follow as discussed offline.

File: notebooks/README.md (hunk 21 above)
Comment (wphicks) [reply notebooks/README.md:29]: Also have them name the container at launch so they can `cp` directly to that name

File: notebooks/simple_xgboost_example.ipynb [hunk 22]
```diff
@@ -0,0 +1,752 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Run the Triton Inference Server in a Subprocess from Jupyter notebook\n",
+    "\n",
+    "# The os.setsid() is passed in the argument preexec_fn so\n",
+    "# it's run after the fork() and before  exec() to run the shell.\n",
+    "triton_process = subprocess.Popen([\"tritonserver --model-repository=/model_repository\"], stdout=subprocess.PIPE, \n",
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:369]:
```suggestion
    "triton_process = subprocess.Popen([\"tritonserver\", \"--model-repository=/model_repository\"], stdout=subprocess.PIPE, \n",
```

File: notebooks/simple_xgboost_example.ipynb [hunk 23]
```diff
@@ -0,0 +1,752 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Run the Triton Inference Server in a Subprocess from Jupyter notebook\n",
+    "\n",
+    "# The os.setsid() is passed in the argument preexec_fn so\n",
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:367]: Are these comments relevant to the notebook?

File: notebooks/simple_xgboost_example.ipynb (hunk 22 above)
Comment (wphicks) [reply notebooks/simple_xgboost_example.ipynb:369]: Taking care of that should also allow you to eliminate the `shell=True` invocation

File: notebooks/simple_xgboost_example.ipynb [hunk 24]
```diff
@@ -0,0 +1,752 @@
...
...Fn/htmBLqD/+B63A9lwAokfzPHLRvHYb5q39HlLIWbJ79irmr/iux2lzmpa/Qw78dRnzwNwxCp2Niuxxsm9EfLYMmYF9CTZkkKZFw6y5SVbNI2NRvAEsLe/XfZcu4hM96tUS74e/jsG4fzJjUHs+2z0PfFq0x+c94KCFAIhFDlN/3CmKIJRJIRAKgxXfY1H0Xbve+6xDsdBG97TMsWbIEnyybjCHjtiHJxBTpl/7E/otPtMgX7fOYvakkMDC+jz+WvI+FCxdidsRorLymh1p69/Hn2sUIbx+Ed7c/hBKAnk7++1IyyPIX3JTJIIOqPXkaex0ox3aMvdpycerwWTxTAhDEcPb0QIkxtcQbU35ai2EN1dx7UCbi1p38C5g2qNfwdZ95aZB7GodOPAEBEAmN4OlZyoIXYleM/X4T3nFVv/6DIvE27qSqGhmJbT00MK6C9L5ORLowMjOHmYlhGe+oimFWzxPNWrVCC58KTCINa6Fxi0D4u1hDF4BILIGuri50dXUhkYgB5OCvacEIWx8PGzcf+HrUgcHjgzh05RkgskKnt7vCpsiQ9gXTw15tL/uWMVOvsrNIkbCWOjkMpZ1Z6XTgPQ/VOxPiRjTlSDql7hpDrsGf0k3pHVrR1rLYo8gKure2L5mLQBC70IzjeY89HZ5CDcUCQWROfdY/IJLH0MeBlqpHrGpF0P4sJRFl08GJXqpj5T+mKr9DX3RUvZeh7zOfLsuURDkHaFRdA4KgQ41nnFQ9QvWKPooMiR/N3PUvHdgyh9rV0iUBAunV70/f38wl9d9l+5H+/aIrGQsgSHzo/YtSIsqhve+6kBgC6btNp9NSJZH8On0caKF69KzD1xSf/y5IZXwXTvp83yLj9vR1bC4RKShh1wc088d/NeeLPFq7PK6yfGAVVa1xzvqZ+hjrECCQSfBnqveZpMdooosxAQLpe86iszIlZf87g5x1BAJ0KGDRRVWZPjKNHMWqtQD0veYQEWm9XXXg8vp6qpZ8VSTQ6s7WeetcmFLYz880/6bQo8j6ftNo97EDtHl2CNmLBYKgS479N9CtvFcT39RHkRUJa6lj3nvI5XmF6vmjyDrUbPpOOnbgB5rZvi6JBZCgW58GbryhWofkdX8UuUaT0omPupCTnQN5tetJfXu1p8Z2hmTWsDWN/uY/SnmJr3JxHlU/fhT5jWWE4LFD0WTjfJz
```
Comment (wphicks) [inline notebooks/simple_xgboost_example.ipynb:645]: It looks like these comments are copied and pasted from above. Are they relevant here?

**Review by wphicks — CHANGES_REQUESTED** [review]
Please review all previous comments and make sure they are addressed.

File: .gitignore [hunk 25]
```diff
@@ -4,3 +4,4 @@ ops/stage
 qa/L0_e2e/model_repository
 qa/L0_e2e/cpu_model_repository
 qa/logs
+*.ipynb_checkpoints*
```
Comment (wphicks) [inline .gitignore:7]:
```suggestion
*.ipynb_checkpoints/*
```

File: notebooks/README.md [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+  --net=host \
+  --name triton_fil
+  nvcr.io/nvidia/tritonserver:<tag>  # Put the appropriate tag here.  
+```
+
+**Note:** The artifacts created by scripts inside the container are created with root permission. The user on host machine might not be able to modify the artifacts once the container exists. To avoid this issue, copy the notebook `docker cp simple_xgboost_example.ipynb <docker_ID>` and create the artifacts inside the container.
+
+Now open up another terminal and copy the notebook from host into the container as follows:
+```
+export CONTAINERID=$(docker ps -aqf "name=triton_fil")
```
Comment (wphicks) [inline notebooks/README.md:34]: Again, let's just name the container when we invoke it to avoid this additional step

**Review by wphicks — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 25 above)
Comment (wphicks) [reply .gitignore:7]: Nevermind; forge that I suggested this. It's fine as is.

File: notebooks/README.md (hunk 21 above)
Comment (abhisheksawarkar) [reply notebooks/README.md]: Yes, added `--name triton_fil` to the docker run command

File: notebooks/simple_xgboost_example.ipynb (hunk 22 above)
Comment (abhisheksawarkar) [reply notebooks/simple_xgboost_example.ipynb]: Ah sorry about this I missed the comma there and it threw an error when I tried to execute it after removing `shell=True` . Updated the command.

File: notebooks/simple_xgboost_example.ipynb (hunk 23 above)
Comment (abhisheksawarkar) [reply notebooks/simple_xgboost_example.ipynb]: Removed the comments

File: notebooks/simple_xgboost_example.ipynb (hunk 24 above)
Comment (abhisheksawarkar) [reply notebooks/simple_xgboost_example.ipynb]: Is it the same comment as before (ones on top of triton_process)? On GH it shows me the screenshot attachment for this issue

File: notebooks/README.md (hunk 26 above)
Comment (abhisheksawarkar) [reply notebooks/README.md]: Removed the additional step and modified docker cp command: 
`docker cp notebooks/ triton_fil:/`

File: notebooks/simple_xgboost_example.ipynb [hunk 27]
```diff
@@ -0,0 +1,714 @@
...
+    "!pip3 install xgboost"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## XGBoost model\n",
+    "\n",
+    "If you have a pre-trained xgboost model, save it as `xgboost.model` and skip this step. We'll train a XGBoost model on random data in this section "
```
[the remark this answers is not in the document]
Comment (wphicks) [reply notebooks/simple_xgboost_example.ipynb:109]: @abhisheksawarkar This still needs to be addressed

File: notebooks/simple_xgboost_example.ipynb (hunk 5 above)
Comment (wphicks) [reply notebooks/simple_xgboost_example.ipynb:144]: @abhisheksawarkar This still needs to be addressed

File: notebooks/simple_xgboost_example.ipynb (hunk 6 above)
Comment (wphicks) [reply notebooks/simple_xgboost_example.ipynb:166]: @abhisheksawarkar This still needs to be addressed

File: notebooks/simple_xgboost_example.ipynb (hunk 27 above)
[the remark this answers is not in the document]
Comment (wphicks) [reply notebooks/simple_xgboost_example.ipynb:109]: (More specifically the first comment on serialization)

File: notebooks/simple_xgboost_example.ipynb (hunk 11 above)
Comment (wphicks) [reply notebooks/simple_xgboost_example.ipynb:344]: @abhisheksawarkar This still needs to be addressed
